draft-irtf-cfrg-vdaf.md

```diff
@@ -3542,60 +3587,151 @@ def unshard(Poplar1, agg_param,
-        input_shares.append(encoded)
-    return input_shares
-
-def decode_input_share(Poplar1, encoded):
-    l = Poplar1.Idpf.KEY_SIZE
-    key, encoded = encoded[:l], encoded[l:]
-    l = Poplar1.Prg.SEED_SIZE
-    corr_seed, encoded = encoded[:l], encoded[l:]
-    l = Poplar1.Idpf.FieldInner.ENCODED_SIZE \
-        * 2 * (Poplar1.Idpf.BITS - 1)
-    encoded_corr_inner, encoded = encoded[:l], encoded[l:]
-    corr_inner = Poplar1.Idpf.FieldInner.decode_vec(
-        encoded_corr_inner)
-    l = Poplar1.Idpf.FieldLeaf.ENCODED_SIZE * 2
-    encoded_corr_leaf, encoded = encoded[:l], encoded[l:]
-    corr_leaf = Poplar1.Idpf.FieldLeaf.decode_vec(
-        encoded_corr_leaf)
-    if len(encoded) != 0:
-        raise ERR_INPUT
-    return (key, corr_seed, corr_inner, corr_leaf)
+uint8 Poplar1FieldInner[Fi];
+~~~
+
+Likewise, elements of the leaf field are encoded in little-endian byte order
+(as defined in {{field}}) and are represented as follows:
+
+~~~
+uint8 Poplar1FieldLeaf[Fl];
+~~~
+
+#### Public Share
+
+The public share is equal to the IDPF public share, which is a byte string.
+(See {{idpf}}.)
+
+#### Input Share
+
+Each input share is structured as follows:
+
+~~~
+struct {
+    uint8 idpf_key[Poplar1.Idpf.KEY_SIZE];
+    Poplar1Seed corr_seed;
+    Poplar1FieldInner corr_inner[Fi * 2 * (Poplar1.Idpf.BITS - 1)];
+    Poplar1FieldLeaf corr_leaf[Fl * 2];
+} Poplar1InputShare;
+~~~
+
+#### Prep Share
+
+Encoding of the prep share depends on the round of sketching: if the first
+round, then each sketch share has three field elements; if the second round,
+then each sketch share has one field element. The field that is used depends on
+the level of the IDPF tree specified by the aggregation parameter, either the
+inner field or the leaf field.
+
+For the first round and inner field:
+
+~~~
+struct {
+    Poplar1FieldInner sketch_share[Fi * 3];
+} Poplar1PrepShareRoundOneInner;
+~~~
+
+For the first round and leaf field:
+
+~~~
+struct {
+    Poplar1FieldLeaf sketch_share[Fl * 3];
+} Poplar1PrepShareRoundOneLeaf;
+
+For the second round and inner field:
+~~~
+~~~
+struct {
+    Poplar1FieldInner sketch_share;
+} Poplar1PrepShareRoundTwoInner;
+~~~
+
+For the second round and leaf field:
+
+~~~
+struct {
+    Poplar1FieldLeaf sketch_share;
+} Poplar1PrepShareRoundTwoLeaf;
+~~~
+
+#### Prep Message
+
+Likewise, the structure of the prep message for Poplar1 depends on the
+sketching round and field. For the first round and inner field:
+
+~~~
+struct {
+    Poplar1FieldInner[Fi * 3];
+} Poplar1PrepMessageRoundOneInner;
+~~~
+
+For the first round and leaf field:
+
+~~~
+struct {
+    Poplar1FieldLeaf sketch[Fl * 3];
+} Poplar1PrepMessageRoundOneLeaf;
+
+Note that these messages are identical to the prep shares for the same round.
+
+The second-round prep message is the empty string. This is because the sketch
+shares are expected to sum to a particular value if the output shares are
+valid; we represent this value by the empty string.
```

Comment:
These paragraphs seem a little confusing, here's a suggested clarification.
```suggestion
Note that these messages have the same structures as the prep shares for the
first round.

The second-round prep message is the empty string. This is because the sketch
shares are expected to sum to a particular value if the output shares are
valid; we represent a successful preparation with the empty string and
otherwise return an error.
```